woocommerce/woocommerce-ios
# [Mobile Payments] Reuse payment intent when retrying

## Description


Previously, whenever a payment was retried, we would attempt to cancel the existing payment intent, and then start the payment again.

This led to most retry attempts failing, and where they worked, added risk that a payment may be charged twice.

This PR changes behaviour to retry with the same payment intent, when that's appropriate, and only use the existing retry handling if the payment intent has not yet been created.

More context available here: p91TBi-520-p2

## Testing instructions


For all of these scenarios, set up Proxyman or Charles to observe your network traffic, with SSL proxying enabled for `public-api.wordpress.com` and `api.stripe.com`.

### Retry Payment Intent creation error

Failures which happen before we get the PI in hand can be retried by restarting. We attempt to cancel any PI that we do have (more relevant to other ways of triggering a restart error, but I don't have any steps), and then create one again.

Set a breakpoint in Proxyman/Charles on `https://api.stripe.com/v1/payment_intents/` for this test. Request or response is fine.

1. Create an order and start a card payment, using either TTP or a card reader
2. The breakpoint will trigger in Proxyman/Charles – abort it, or wait 30s for it to time out
3. Observe that a retryable error is shown
4. Tap `Try Collecting Again`
5. Tap the testing card 
6. Observe that the payment is completed
7. Look at Proxyman/Charles – observe that the two POST calls to `https://api.stripe.com/v1/payment_intents/`, and that the ID returned in the second one was used for the subsequent payments.

### Retry card tap

Failures that happen often require the card to be tapped again, in particular, errors which are resolved by using a different card.

1. Create an order and start a card payment, using either TTP or a card reader
2. Tap a live card
3. Observe that a retryable error is shown saying that you can't use a live card in test mode
4. Tap `Try Collecting Again`
5. Tap the testing card 
6. Observe that the payment is completed
7. Look at Proxyman/Charles – observe that the two POST calls to `https://api.stripe.com/v1/payment_intents/{id}/confirm` show the same ID was used. The first one will have failed, the second will have succeeded.

### Retry processing

Network errors can interrupt processing, which can then be retried without needing to capture card details again.

Set a breakpoint in Proxyman/Charles on **requests to** `https://api.stripe.com/v1/payment_intents/*/confirm` for this test. The response is too late, as Stripe will have processed the payment.

1. Create an order and start a card payment, using either TTP or a card reader
2. Tap a test card
3. Abort the request that comes in to Proxyman/Charles at your breakpoint.
4. Observe that a retryable error is shown saying that your internet connection is not available
5. Tap `Try Collecting Again`
6. Allow the breakpoint to continue
7. Observe that the payment is completed, and you are not asked to tap the card again.
8. Look at Proxyman/Charles – observe that the two POST calls to `https://api.stripe.com/v1/payment_intents/{id}/` show the same ID was used. The first one will have failed, the second will have succeeded.

### Retry capture

Set a breakpoint in Proxyman/Charles on **responses from** `https://api.stripe.com/v1/payment_intents/*/confirm` for this test.

1. Create an order and start a card payment, using either TTP or a card reader
2. Tap a test card
3. Abort the request that comes in to Proxyman/Charles at your breakpoint.
4. Observe that a retryable error is shown saying that your internet connection is not available
5. Tap `Try Collecting Again`
6. Allow the breakpoint to continue
7. Observe that the payment is completed, and you are not asked to tap the card again.
8. Look at Proxyman/Charles – observe that the two POST calls to `https://api.stripe.com/v1/payment_intents/{id}/` show the same ID was used. Both will show as failures, but the second one will have the message below:

```
"message": "You cannot provide a new payment method to a PaymentIntent when it has a status of requires_capture. You should either capture the remaining funds, or cancel this PaymentIntent and try again with a new PaymentIntent."
```

Note that the mechanism here is different – when we get the second error, we treat that as successful processing, and go on to capture the payment via WCPay as normal

---
